serialization/class_definition.go

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package serialization
+
+import "github.com/hazelcast/hazelcast-go-client/hzerrors"
+
+// ClassDefinition defines a class schema for Portable structs.
+type ClassDefinition struct {
+	Fields    map[string]FieldDefinition
+	FactoryID int32
+	ClassID   int32
+	Version   int32
+}
+
+func NewClassDefinition(factoryID int32, classID int32, version int32) *ClassDefinition {
+	return &ClassDefinition{
+		FactoryID: factoryID,
+		ClassID:   classID,
+		Version:   version,
+		Fields:    make(map[string]FieldDefinition),
+	}
+}
+
+func (cd *ClassDefinition) AddField(definition FieldDefinition) {
+	cd.Fields[definition.Name] = definition
+}
+
+func (cd *ClassDefinition) AddByteField(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeByte)
+}
+
+func (cd *ClassDefinition) AddBoolField(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeBool)
+}
+
+func (cd *ClassDefinition) AddUInt16Field(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeUint16)
+}
+
+func (cd *ClassDefinition) AddInt16Field(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeInt16)
+}
+
+func (cd *ClassDefinition) AddInt32Field(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeInt32)
+}
+
+func (cd *ClassDefinition) AddInt64Field(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeInt64)
+}
+
+func (cd *ClassDefinition) AddFloat32Field(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeFloat32)
+}
+
+func (cd *ClassDefinition) AddFloat64Field(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeFloat64)
+}
+
+func (cd *ClassDefinition) AddStringField(fieldName string) {
+	cd.addNewFieldDefinition(fieldName, TypeString)
+}
+
+func (cd *ClassDefinition) AddPortableField(fieldName string, def *ClassDefinition) error {
+	if def.ClassID == 0 {
+		return hzerrors.NewHazelcastIllegalArgumentError("Portable class ID cannot be zero", nil)
+	}
+	cd.AddField(NewFieldDefinition(int32(len(cd.Fields)), fieldName, TypePortable, def.FactoryID, def.ClassID, cd.Version))
+	return nil
+}
+
+func (cd *ClassDefinition) addNewFieldDefinition(fieldName string, fieldType int32) {
+	cd.AddField(NewFieldDefinition(int32(len(cd.Fields)), fieldName, fieldType, 0, 0, cd.Version))
```

Comment:
In Java we do not permit to use of duplicate field names

Reply:
The same here, the added field ends up at: `cd.Fields[definition.Name] = definition`.

Reply:
But it is a bit different here. So, in Java if someone does this, it will fail.
```java
new ClassDefinitionBuilder(1, 1,1)
              .addBooleanField("a")
              .addIntField("a")
              .build();
```

If I am not missing something, with this API, the following is possible in Go
```go

cd := NewClassDefinition(1, 1, 1)
cd.AddBooleanField("a")
cd.AddInt32Field("a")
```

Reply:
OK, I've added a check for that and return `error`s from AddXXX fields :(( https://github.com/hazelcast/hazelcast-go-client/pull/569/commits/3c294702ef9bc90212e7d5e65fe10896a80e757d